### PR TITLE
fix: ensure has() does not change v:shell_error

### LINKED ISSUE #18278: `has('python3')` call leaves `v:shell_error` set
### Neovim version (nvim -v)

55135cea619cd1b8b8d7563c14436c092fa749ab

### Vim (not Nvim) behaves the same?

No

### Operating system/version

macOS 12

### Terminal name/version

iTerm2

### $TERM environment variable

N/A

### Installation

From source

### How to reproduce the issue

Run `nvim --clean` on a system that does not have the `neovim` Python 3 package installed, then:

```
:call has('python3')
:echom v:shell_error
```

### Expected behavior

The `has('python3')` call returns false, and `v:shell_error` is either falsy (0) or unchanged from before the `has(...)` call (this is how Vim behaves, the value of `v:shell_error` is unchanged by calls to `has`).

### Actual behavior

The `has('python3')` call returns false, and `v:shell_error` is set to 2.

I ran into this because if `:PlugUpdate my-package` is called where `my-package` is pinned to a particular commit, and `my-package` is already checked out to that commit, then `vim-plug`'s code happens not to invoke any other shell commands between a `has('python3')` call ([here](https://github.com/junegunn/vim-plug/blob/be55ec46b5ab031805487a9c3e895db57f53010b/plug.vim#L1192)) and a closing block where it checks whether something went wrong by inspecting `v:shell_error` (somewhere in [here](https://github.com/junegunn/vim-plug/blob/be55ec46b5ab031805487a9c3e895db57f53010b/plug.vim#L1284), I think). (So `vim-plug` reports a spurious error.)

So either
- we need to clear `v:shell_error` at some point in the implementation of `has(...)` -- I looked at this a bit and `v:shell_error` is read-only so I'm not sure how to proceed, or
- if this is "working as intended" then we can close this issue and I can refile as a `vim-plug` bug.

### PR DESCRIPTION
**Pull request by andmis**
Resolves #18278 .

### CODE REVIEW

File: src/nvim/eval/funcs.c [hunk 1]
```diff
@@ -4278,6 +4278,7 @@ static void f_has(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     "nvim",
   };
 
+  const int original_shell_error = get_vim_var_nr(VV_SHELL_ERROR);
```
Comment (justinmk) [inline src/nvim/eval/funcs.c:4281]: - let's add a small comment
- codebase convention is `save_foo`

```suggestion
  // XXX: eval_has_provider() may shell out :(
  const int save_shell_error = get_vim_var_nr(VV_SHELL_ERROR);
```

**Comment by justinmk** [comment]
Thanks!  Should be easy to add a test to `test/functional/vimscript/has_spec.lua`

You can [run the test](https://github.com/neovim/neovim/blob/master/test/README.md#running-tests) like this:

    TEST_FILE=test/functional/example_spec.lua make functionaltest

**Comment by andmis** [comment]
Thanks for the pointers. Changes incorporated.

File: test/functional/vimscript/has_spec.lua [hunk 2]
```diff
@@ -68,4 +68,10 @@ describe('has()', function()
       eq(0, funcs.has('wsl'))
     end
   end)
+
+  it('does not change v:shell_error', function()
+    funcs.system('bash -c "exit 73"')
```
Comment (justinmk) [inline test/functional/vimscript/has_spec.lua:73]: these tests run on windows. a platform-agnostic way to force an exit code is to run `nvim` with `:cquit`:

```lua
local nvim_prog = helpers.nvim_prog

funcs.system({nvim_prog, '+73cquit'})
```

**Review by justinmk — APPROVED** [review]
LGTM after last comment

File: test/functional/vimscript/has_spec.lua (hunk 2 above)
Comment (andmis) [reply test/functional/vimscript/has_spec.lua]: Ah. Thanks.

File: test/functional/vimscript/has_spec.lua [hunk 3]
```diff
@@ -68,4 +68,11 @@ describe('has()', function()
       eq(0, funcs.has('wsl'))
     end
   end)
+
+  it('does not change v:shell_error', function()
+    local nvim_prog = helpers.nvim_prog
+    funcs.system({nvim_prog, '+73cquit'})
```
Comment (justinmk) [inline test/functional/vimscript/has_spec.lua:74]:
```suggestion
    funcs.system({nvim_prog, '-es', '+73cquit'})
```

**Comment by andmis** [comment]
@justinmk TY for shepherding this through.

**Comment by clason** [comment]
candidate for backport?

**Pull request merged**